### PR TITLE
Add memory system call replacements.

### PR DESCRIPTION
**Pull request by davidchisnall**
These system calls are used for heap setup and so we can't yet remove
the musl wrappers, but this provides the kernel support that we'll need
when we do.

### CODE REVIEW

File: src/enclave/enclave_mem.c
Comment (prp) [inline src/enclave/enclave_mem.c]: Is there some weird indentation here?

File: src/include/enclave/enclave_mem.h
Comment (prp) [inline src/include/enclave/enclave_mem.h:27]: Why not define `mmap_files` here?

**Review by prp — APPROVED** [review]
LGTM

File: src/include/enclave/enclave_mem.h
Comment (davidchisnall) [reply src/include/enclave/enclave_mem.h:27]: And make it common linkage?  I try to avoid common linkage whenever possible because it causes all manner of subtle linker behaviours.

File: src/enclave/enclave_mem.c
Comment (davidchisnall) [reply src/enclave/enclave_mem.c]: I forgot to run clang-format before I committed this.  Fixed now.

**Review by vtikoo — APPROVED** [review]

File: src/enclave/enclave_mem.c
Comment (SeanTAllen) [inline src/enclave/enclave_mem.c:130]: Did you intend to remove this lseek when this function was moved and partially rewritten?

Comment (davidchisnall) [reply src/enclave/enclave_mem.c:130]: Yes.  lseek changes the offset in the fd, which is something mmap should not do.

Comment (SeanTAllen) [reply src/enclave/enclave_mem.c:130]: David replied elsewhere that it was intentional

**Comment by SeanTAllen** [comment]
I'm working on this PR now.

**Comment by SeanTAllen** [comment]
This is currently blocked waiting for #577 to land. In the meantime, I've done some related work in #655.

Once #655 and #577 are merged to oe_port, I'm going to pick up the work on this again and remove the syscall overrides for mmap functions from sgx-lkl-musl and into system call replacements like this PR has.

Some of the changes needed to make this change possible are part of #655.

**Comment by SeanTAllen** [comment]
Closed in favor of #710. I'm leaving the branch around until #710 is merged.

**Pull request closed without merging**
